---
title: "CITEseq data set generation"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CITEseq data set generation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# CITEseq data set

This document describes how we generated the citeseq data set used in the
software package `AbNames` and the data package `CITEseq`.  This data set
is a curated collection of antibody names from publicly available CITE-seq
data sets.  Generating this data set involved several steps:

* Collecting antibody information
* Matching antibody information to Antibody Derived Tag (ADT) names in
expression tables.
* Correcting errors by comparison between data sets
* Annotating antibody information with gene identifiers
* Generating standardized antibody names

We discuss all the steps in detail below.

## Collecting antibody information

We first created a master table of antibody vendor information, including
antibody names, vendors, vendor catalogue numbers and antibody clones. For some
studies, the antibody clone table was available as a supplementary table.  For
others, we manually created clone tables from the information in the methods or
resources section.  When the clone information was available in a downloadable
format,  we have manually fixed some entries that were incorrectly formatted due
to file encoding issues.  We have also converted Greek symbols to letters.
We rename columns from individual antibody clone tables for consistency between
experiments. 

```{r}
# This script generates a combined ADT clone table
source("data-raw/01_unify_adt_clones.R")
ls()
head(all_clones)
```

### Correcting errors by comparison between data sets

## Matching antibody clone names with data names

The directory `inst/extdata/protein_names` contains the feature names from the
antibody derived tag expression matrices.  These files are named according to
the study accessions.

## Correcting errors by comparison between data sets

Especially as some of the antibody tables were generated by copy-pasting,
we discovered some inconsistencies in antibody vendor information between
studies.  We have correctly entries where the antibody clone, catalogue number
or oligo ID did not match the information provided by the vendor.
We identified such errors by combining the clone
tables for all studies, grouping by clone, catalog number, antigen or
oligo ID, and looking for cases where there were multiple values in one
of the other columns, e.g. one catalogue number should match exactly one
oligo ID.  When there was a discrepancy, we checked the vendor website and
the original manuscript to determine the (likely) true value.  Tables of the corrections made are found in inst/extdata/01_unify_adt_clones.R.


# Notes 

HGNC IDs are stable, HGNC symbol can change
UniProt assigns isoforms an id composed on primary accession plus - number

NCI thesaurus
```{r}
paste(c("https://www.ebi.ac.uk/ols/ontologies/ncit/",
       "terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FNCIT_C118492"),
     collapse = "")
```

Corum downloadable mammalian protein complexes
Not useful, requires free text search
http://mips.helmholtz-muenchen.de/corum/

Downloadable cell marker database (but not protein complexes)
ttp://bio-bigdata.hrbmu.edu.cn/CellMarker/download.jsp

Cell surface protein atlas:
http://wlab.ethz.ch/cspa/#news

R package clustermole:
```{r}
clustermole <- paste(c("https://cran.r-project.org/web/packages/",
                      "clustermole/vignettes/clustermole-intro.html"),
                    collapse = "")
```

Package rols offers protein ontology lookup with non-exact matching  


After filtering, only KLRG1 / MAFA maps to different genes, decide manually
Uniprot lists MAFA as a synonym for KLRG1, also MAFA-like
From BioLegend:
KLRG1 is the mouse homologue of rat MAFA, 
Human homologue is expressed on subsets of NK cells, 
CD4+ and CD8+ effector/memory cells
In org.db - MAFA is an alias for KLRG1, but MAFA as a symbol is a 
different gene with different name

CD20 maps to KRT20, MS4A1, only MS4A1 is correct
CD158b mabs to two symbols, many gene IDs
CD98 is a heterodimer of SLC3A2 and SLC7A5, both correct
Not sure if KIR2DL5 should be both KIR2DL5B and KIR2DL5A
From: https://www.frontiersin.org/articles/10.3389/fimmu.2012.00289/full
KIR2DL5 alleles belong to two series encoded by different loci 
...designated officially KIR2DL5A and KIR2DL5B
CD277 - not sure if it's both BTN3A1 and BTN3A2
anti-PE - 
If it's an Ig[AMGD] don't trust the toupper symbol
- don't want putative locations?
GP130 = IL6ST = CD130? Can't tell from org.db table
TCRb and TCRg match with uppercase, but multiple genes
TCR / CD3 - which part(s) do the antibodies recognise?

If replacing greek letters, be careful esp with "mu"!

TO FIX: 
ROWS GAINED AT SOME POINT, BECAUSE OF TRZUPEK DUPLICATES?  
SYMBOLS SHOULD BE UPPER, E.G. CD8A, BUT CAREFUL WITH IG
IG GENES HAVE PREFIXES IGH IGL, IGK - lambda kappa heavy?
TCR genes are TRA/B/D for TCR-alpha/beta/delta
WHY ARE HUMAN AND MOUSE IN THE TABLE?
IS B7-RP1 = RP1? - No!
CHECK IF THERE IS A MATCH TO THE FULL PROTEIN, ONLY USE SUBUNIT IF THERE IS 
NOT

Does merging in manual matches create duplicate rows?

# Lower no dash examples:

```{r}
l_no_dash <- structure(list(greek_letter =
    c("Galectin-3", "Podoplanin", "Nectin-2", 
      "Siglec-7", "Thrombomodulin", "Notch 3", "Sialoadhesin", "Siglec-1", 
      "Notch 1", "Langerin", "Syndecan-1", "Neuropilin-1", "Cadherin 11", 
      "Podocalyxin", "Trail", "Adenosine", "Galectin-9", "Phospho", 
      "Siglec-8")),
    class = "data.frame", row.names = c(NA, -19L))
```

Match short names with CITEseq package.
